jgrapht-core/src/main/java/org/jgrapht/graph/concurrent/AsSynchronizedGraph.java

```diff
@@ -0,0 +1,1405 @@
+        public boolean removeEdge(E e)
+        {
+            V sourceVertex = getEdgeSource(e);
+            V targetVertex = getEdgeTarget(e);
+            if (AsSynchronizedGraph.super.removeEdge(e)) {
+                edgeModified(sourceVertex, targetVertex);
+                return true;
+            }
+            return false;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public E removeEdge(V sourceVertex, V targetVertex)
+        {
+            E e = AsSynchronizedGraph.super.removeEdge(sourceVertex, targetVertex);
+            if (e != null)
+                edgeModified(sourceVertex, targetVertex);
+            return e;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public boolean removeVertex(V v)
+        {
+            if (AsSynchronizedGraph.super.removeVertex(v)) {
+                edgesOfMap.clear();
+                incomingEdgesMap.clear();
+                outgoingEdgesMap.clear();
+                return true;
+            }
+            return false;
+        }
+
+        /**
+         * Clear the copies which the edge to be added or removed can affect.
+         *
+         * @param sourceVertex source vertex of the modified edge.
+         * @param targetVertex target vertex of the modified edge.
+         */
+        private void edgeModified(V sourceVertex, V targetVertex)
+        {
+            outgoingEdgesMap.remove(sourceVertex);
+            incomingEdgesMap.remove(targetVertex);
+            edgesOfMap.remove(sourceVertex);
+            edgesOfMap.remove(targetVertex);
+            if (!AsSynchronizedGraph.super.getType().isDirected()) {
+                outgoingEdgesMap.remove(targetVertex);
+                incomingEdgesMap.remove(sourceVertex);
+            }
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public boolean isCacheEnabled()
+        {
+            return true;
+        }
+    }
+
+    /**
+     * A builder for {@link AsSynchronizedGraph}.
+     *
+     * @author CHEN Kui
+     */
+    public static class Builder
+    {
+        private boolean cacheEnable;
+        private boolean fair;
+
+        /**
+         * Construct a new Builder with cacheDisable and non-fair mode.
+         */
+        public Builder()
+        {
+            cacheEnable = false;
+            fair = false;
+        }
+
+        /**
+         * Construct a new Builder.
+         *
+         * @param graph the graph to base the builder
+         */
+        public Builder(AsSynchronizedGraph graph)
+        {
+            this.cacheEnable = graph.isCacheEnabled();
+            this.fair = graph.isFair();
+        }
+
+        /**
+         * Request a synchronized graph using non-fair mode without caching.
+         *
+         * @return the Builder
+         */
+        public Builder cacheDisable()
+        {
+            cacheEnable = false;
+            fair = false;
```

Comment:
Why does this method affect fair?  It should be limited to cache.

Reply:
I'am so sorry for this bug